### ISSUE TITLE
Squeezer intermittently loses connection to server

### ISSUE THREAD

**Issue by grassy-noel** [nikclayton/android-squeezer#177]
If I have Squeezer open and continually use the UI, everything works fine.
However, if I switch to another app on my tablet like GMail for a period of time and switch back, Squeezer will hang trying to re-connect to the Squeezebox server.

BTW - I cloned this repo onto my machine to try to Debug the issue myself. But there's a ton of dependencies missing from this repo and a ton of errors trying to get this compiling in Android studio.

I'd love to be able to contribute a fix.

**Comment by matt-matt2** [comment]
I'm seeing the same issue, in that it will only successfully connect to LMS on a cold start so I need to kill the app and restart.

**Comment by matt-matt2** [comment]
FWIW if I had to guess I'd say this started happening when my android got updated to 6.0. also happens on android 7.0

**Comment by kaaholst** [comment]
If you switch to the "develop" branch I think you will more successful compiling in Android Studio.

I am not able to reproduce this issue on my devices.

**Comment by grassy-noel** [comment]
Some further feedback. I see this on multiple Android devices. But I know it's happening on an Lenovo A8-50 tablet running Lollipop 5.04.

I've also pulled down the develop branch as suggested.. Most of the required libs are missing from this branch. I copied them in from master.  After about 3 hours of fighting with Gradle, I still can't get Android studio to output an apk for debugging. Mainly becuase of the depencies the app has on Crashylitics and Changelog which aren't in the Github repo.

**Comment by kaaholst** [comment]
HHmm...
Dependencies are supposed to to be downloaded automatically by gradle, they are not supposed to be in the github repo.

The dependencies section of Squeezer/build.gradle lists the dependencies.

About cloning, this is where our repo lives: https://github.com/nikclayton/android-squeezer.
There is README.md which tells about the project, and a CONTRIBUTING.md with guidelines about how to contribute.

**Comment by grassy-noel** [comment]
ok, so pulled down the develop branch and along with all of the other changes I'd made to Android studio and Gradle along the way, I got it to produce an APK.
Thing is, the develop branch is building version 1.4 not the version 1.6 that is on Google Play.
Also version 1.4 doesn't seen to have the sleep/resume issue that 1.6 has.

**Comment by matt-matt2** [comment]
Hmm, I've got v1.4.1 from Google Play.
-- 
Sent from wherever I am.

**Comment by nikclayton** [comment]
I recently merged 1.4.1 back in to the develop branch (oversight on my part when I cut the last release) so you should be able to see it now.

FWIW I also occasionally see this behaviour, but have been unable to reliably reproduce it, which makes fixing it somewhat tricky unfortunately. Have you ever been able to trigger it repeatedly?

**Comment by matt-matt2** [comment]
I started seeing it when Android 6 hit my devices and can reproduce it almost 100%. You need to force the app to suspend whilst in the background, i.e. run enough other apps that the OS forces it to due to shortage of RAM. With devices of different specification it's likely the amount of other apps you have to run will be very different. My phone has 2G ram and any of the "big" apps (Facebook, FB Messenger, Hangouts, Firefox, Chrome), will generally force backgrounded apps to suspend and trigger the problem.

**Comment by nikclayton** [comment]
Yeah, I've been trying that, but with no joy.

I've also tried force-killing. Specifically:

1. Run Squeezer.
2. Go to, e.g. the "Now Playing" screen.
3. Press the "Home" button to get out of Squeezer
4. Kill it with `adb shell am force-stop uk.org.ngo.squeezer`
5. Get back to it by using the "Recent apps" button and selecting it from there.

I see the "Connecting" dialog popup briefly, and then it's back to where it was before. On a physical device running 6.0.1.

Looking at what changed between 1.4.0 and 1.4.1 there's not much that jumps out at me.

It's a bit of a reach, but what do you have the Squeezer "Notification" configuration option set to?

**Comment by matt-matt2** [comment]
No, I think that's the equivalent of my "swipe to kill" in the task switcher dialogue, which seems to leave a clean state and the app starts up again ok. The issue is that android has different levels of sleep, from essentially "please be quiet" through "save state and go to sleep" to "terminate with prejudice". I think since android 6 came in it is taking the first couple of levels of suspend more aggressively and that's what is not being handled right.

-- 
Sent from wherever I am.

**Comment by matt-matt2** [comment]
Doze mode came in 6.0. could this be related? https://developer.android.com/training/monitoring-device-state/doze-standby.html

**Comment by nikclayton** [comment]
Oh, that's looks relevant. Especially the text about network connections being dropped. When it's running Squeezer locks the WiFi so it doesn't have to deal with the network disappearing, and it looks like Doze overrides that.

While I investigate a code fix, can you try and exempting Squeezer from Doze (Android Settings > Battery > Battery Optimization, find Squeezer, set the "Don't optimise" option) and see if that makes the problem go away for you. That'll confirm if this is the root cause.

**Comment by matt-matt2** [comment]
Certainly if it assumes that existing sockets can be reused without an explicit reconnect that would explain why it can't restart. I'll check that when I get home.

I would guess the fix is to detect disconnection sooner and loop back round to a full reconnect?
-- 
Sent from wherever I am.

**Comment by matt-matt2** [comment]
Can confirm that after disallowing battery optimise for squeezer I can no longer reproduce the problem. If it gets shut down now after being backgrounded it starts up cleanly every time.


On 26 February 2017 12:57:24 GMT+00:00, Nik Clayton <notifications@github.com> wrote
>While I investigate a code fix, can you try and exempting Squeezer from
>Doze (Android Settings > Battery > Battery Optimization, find Squeezer,
>set the "Don't optimise" option) and see if that makes the problem go
>away for you. That'll confirm if this is the root cause.

-- 
Sent from wherever I am.

**Comment by cwarden** [comment]
I'm still able to reproduce after disabling battery optimization for Squeezer 1.4.1.  I'm running Android 7.1.1 on a Nexus 5X.

**Comment by matt-matt2** [comment]
It happened to me this morning too, but only after a longer period asleep (overnight). It's much more stable in day to day use.
-- 
Sent from wherever I am.

**Comment by cwarden** [comment]
My experience is similar to @matt-matt2's.  I was unable to reproduce it yesterday, but trying to resume the app this morning, I hit the bug.

**Comment by nikclayton** [comment]
Managed to reproduce this in the emulator. For the record:

Make sure Doze is enabled and emulator thinks the device is on battery power.

```
adb shell dumpsys deviceidle enable
adb shell dumpsys battery unplug
```

Send keyevent 26, which is a tap of the power button to turn the screen off.

```
adb shell input keyevent 26
```

NowPlayingFragment will correctly pause. The connection to the server is maintained.

Force the device in to Doze mode.

```
adb shell dumpsys device force-idle
```

Network access is suspended. You'd think that would cause this code in CLIClient.java to take the exception path

```
                try {
                    line = in.readLine();
                } catch (IOException e) {
                    line = null;
                    exception = e;
                }
```

which would trigger the normal "unexpectedly disconnected from the server" logic, but no.

Now I can reproduce this I should be able to fix it.

**Comment by grassy-noel** [comment]
I'm in the process of trying to build the develop branch on my local machine for testing.
I see there's some specific Marshmallow code in the check in. But I'm pretty certain that my device is on Lollipop 5.0.4.  It's a Lenovo TAB 2 A8-50.
When you switch back to Squeezer from any other app, the SqeezeCenter login dialog appears on the screen, but the connection never succeeds.

**Comment by cwarden** [comment]
I'm getting a crash upon resume running f82bc9f6 on Android 7.1.1

**Comment by matt-matt2** [comment]
Possibly unrelated, but I had a whole bunch of apps crashing on 7.1.1 on
a Nexus 9 yesterday after a google update was installed. It's stable
again after a reboot. As much as it ever is, anyway.

**Comment by grassy-noel** [comment]
Just tested last night on my tablet. It's on Lollipop 5.1 and with 1Gb of RAM it suspends apps put into the foreground when I swapped to my browser or even let it go to sleep for a sustained period.  This code change worked. Line 227 of the squeezeservice.java
I occasionally saw the server connection dialog and had to re-connect manually, but this succeeded, unlike the behavior previously, Most of the time, the app resumed as expected.

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);

                if ((pm.isDeviceIdleMode()) || (pm.isPowerSaveMode())) {
                    Log.d(TAG, "Entering doze mode, disconnecting");
                    disconnect();
                }
            }

**Comment by nikclayton** [comment]
I don't understand how that change would affect a Lollipop device.

Specifically, the code that registers the receiver is also wrapped with a version check in `onCreate()`, 

```
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            registerReceiver(deviceIdleModeReceiver, new IntentFilter(
                    PowerManager.ACTION_DEVICE_IDLE_MODE_CHANGED));
        }
```

so if you didn't modify that as well the receiver won't even be registered on a Lollipop device.

**Comment by kaaholst** [comment]
Fixed in version 1.5.4 which introduces a foreground service to keep the connection to LMS
